src/react.ts

```diff
@@ -20,7 +20,40 @@ export abstract class ReactComponent<P> implements IReactComponent<P> {
     abstract render(): ReactNode;
 }
 
-export declare function reactDOMRender(input: ReactElement, element: Element): void;
+export const setAttributes = (el: HTMLElement, attrs: any) : void => {
+    for(var key in attrs) {
+    el.setAttribute(key, attrs[key]);
+    }
+}
+
+declare function renderCustomComponent<P>(type: IReactComponentConstructor<P>, props: P ): HTMLElement
+
+export function reactDOMRender(input: ReactElement, element: Element): any{
+    let dOMElement: any;
+    if (typeof input === 'string'){
+        // handle text nodes
+        dOMElement = document.createTextNode(input)
+    } else{
+        const {type, props} = input;
+        if(typeof type !== 'string'){
+            // handle custom components
+            dOMElement = renderCustomComponent(type, props)
+        }else{
+            // handle html tags
+            dOMElement = document.createElement(type)
+            if(props){
+                const {children, ...otherProps} = props;
+                setAttributes(dOMElement, otherProps);
+                if(children){
+                    [].concat(children).forEach(child => {
+                        reactDOMRender(child, dOMElement)
```

Comment:
The magic is right there and it looks correct!
I think the logic could be simplified a bit, otherwise looks good.

Interesting moment is that `renderCustomComponent` is going to need to render children too, and those children could be html tags, so how can you refactor this to reuse the logic?

Reply:
I mean what's preventing us from having `renderCustomComponent` call `reactDOMRender` recursively ?

Reply:
Nothing really, as long as it works.

I don't think you can get away with implementing `renderCustomComponent` without either some duplication or refactoring `reactDOMRender`, because `reactDOMRender` expects `renderCustomComponent` to return an element, while the element is currently created inside `reactDOMRender` and not returned.